File: typescript/src/game-manager.ts
import P5 from 'p5';
import { GameState } from './game-states';
import SpritesManager from './sprites/sprites-manager';
import OverlaysManager from './ui/overlays/overlays-manager';
import Interval from './utils/interval';

const LIFES_WHEN_STARTING = 3;
const ASTEROIDS_START_COUNT = 12;
const ASTEROIDS_LEVEL_INCREMENT = 3;
const GAME_OVER_STATE_TIMEOUT = 8000; // ms
const ADD_LIFE_WHEN_SCORED = 3000;
const ASTEROID_HIT_SCORE = 10;
const UFO_HIT_SCORE = 50;
const UFO_CREATE_INIT_FREQUENCY = 15000; // ms
const UFO_CREATE_DECREMENT_FREQUENCY = 1000; // ms
const UFO_CREATE_MINIMAL_FREQUENCY = 10000; // ms
const UFO_SHOOT_INIT_FREQUENCY = 15000; // ms
const UFO_SHOOT_DECREMENT_FREQUENCY = 500; // ms
const UFO_SHOOT_MINIMAL_FREQUENCY = 5000; // ms

export default class GameManager {
    private p5: P5;

    private spritesManager: SpritesManager;
    private overlaysManager: OverlaysManager;

    private gameState = GameState.HOMESCREEN;

    private level: number;
    private lifes: number;
    private score: number;
    private topScore: number = 0;
    private maxAsteroids: number;
    private gameOverInterval = new Interval();
    private lifeAddedSoFar: number;
    private ufoCreateFrequency: number;
    private ufoShootFrequency: number;
    private gamePaused = false;

    constructor(p5: P5) {
        this.p5 = p5;

        this.spritesManager = new SpritesManager(p5);
        this.overlaysManager = new OverlaysManager(p5);

        this.spritesManager.createAsteroids(ASTEROIDS_START_COUNT);
        this.spritesManager.createUfo(0);

        // Get the cookie top score
        for (const cookie of document.cookie.replace(/ /g, '').split(';')) {
            const cookieSplit = cookie.split('=');

            if (cookieSplit[0] === 'top-score') {
                this.topScore = parseInt(cookieSplit[1]);
            }
        }
    }

    update() {
        this.overlaysManager.update();

        if (!this.gamePaused) {
            this.spritesManager.update();

            if (this.gameState === GameState.PLAYING) {
                this.checkLevel();
                this.checkNewLife();
            }
        }

        if (this.gameOverInterval.isElapsed()) {
            this.gameOverInterval.cancel();
            this.gameState = GameState.HOMESCREEN;

            this.spritesManager.reset();
            this.spritesManager.createUfo(0);
            this.spritesManager.createAsteroids(ASTEROIDS_START_COUNT);
        }
    }

    draw() {
        this.overlaysManager.drawBackground();

        this.spritesManager.draw();

        this.overlaysManager.drawForeground(
            this.gameState,
            this.gamePaused,
            this.topScore,
            this.getScore(),
            this.level,
            this.lifes
        );
    }

    keyPressed(keyCode) {
        this.overlaysManager.keyPressed(keyCode);

        switch (this.gameState) {
            case GameState.HOMESCREEN:
            case GameState.GAME_OVER:
                if (keyCode === 83) {
                    // s
                    this.startGame();
                }

                break;
            case GameState.PLAYING:
                this.spritesManager.keyPressed(keyCode);

                if (keyCode === 80) {
                    // p
                    this.gamePaused = !this.gamePaused;

                    if (this.gamePaused) {
                        this.spritesManager.pause();
                    } else {
                        this.spritesManager.unpause();
                    }
                }

                break;
            case GameState.NEXT_LEVEL:
                if (keyCode === 83) {
                    // s
                    this.nextLevel();
                }

                break;
            case GameState.NEXT_LIFE:
                if (keyCode === 83) {
                    // s
                    this.startLevel();
                }

                break;
        }
    }

    keyReleased(keyCode) {
        this.spritesManager.keyReleased(keyCode);
    }

    private getScore() {
        return (
            this.score +
            this.spritesManager.countAsteroidsHit * ASTEROID_HIT_SCORE +
            this.spritesManager.countUfosHit * UFO_HIT_SCORE
        );
    }

    private startGame() {
        this.gameOverInterval.cancel();

        this.score = 0;
        this.level = 1;
        this.maxAsteroids = ASTEROIDS_START_COUNT;
        this.lifes = LIFES_WHEN_STARTING;
        this.overlaysManager.setLifeCount(this.lifes - 1);
        this.lifeAddedSoFar = 0;
        this.ufoCreateFrequency = UFO_CREATE_INIT_FREQUENCY;
        this.ufoShootFrequency = UFO_SHOOT_INIT_FREQUENCY;

        this.spritesManager.reset();

        this.startLevel();
    }

    private startLevel() {
        this.gameState = GameState.PLAYING;

        this.spritesManager.startLevel(
            this.maxAsteroids,
            this.ufoCreateFrequency,
            this.ufoShootFrequency
        );
    }

    private nextLevel() {
        this.level++;
        this.maxAsteroids += ASTEROIDS_LEVEL_INCREMENT;
        this.ufoCreateFrequency -= UFO_CREATE_DECREMENT_FREQUENCY;
        this.ufoShootFrequency -= UFO_SHOOT_DECREMENT_FREQUENCY;

        if (this.ufoCreateFrequency < UFO_CREATE_MINIMAL_FREQUENCY) {
            this.ufoCreateFrequency = UFO_CREATE_MINIMAL_FREQUENCY;
        }

        if (this.ufoShootFrequency < UFO_SHOOT_MINIMAL_FREQUENCY) {
            this.ufoShootFrequency = UFO_SHOOT_MINIMAL_FREQUENCY;
        }

        this.startLevel();
    }

    private checkLevel() {
        if (!this.spritesManager.shipHit()) {
            if (this.spritesManager.getAsteroidsCount() === 0) {
                this.gameState = GameState.NEXT_LEVEL;
                this.score = this.getScore();
                this.spritesManager.stopLevel();
            }
        } else {
            this.lifes--;
            this.overlaysManager.setLifeCount(this.lifes - 1);
            this.score = this.getScore();
            this.spritesManager.stopLevel();

            if (this.lifes === 0) {
                this.gameState = GameState.GAME_OVER;

                this.topScore =
                    this.score > this.topScore ? this.score : this.topScore;

                // Save the top score in a cookie
                document.cookie = `top-score=${this.topScore}`;

                // Return to homescreen after some time...
                this.gameOverInterval.set(GAME_OVER_STATE_TIMEOUT);
            } else {
                this.gameState = GameState.NEXT_LIFE;
            }
        }
    }

    private checkNewLife() {
        if (
            this.getScore() - this.lifeAddedSoFar * ADD_LIFE_WHEN_SCORED >=
            ADD_LIFE_WHEN_SCORED
        ) {
            this.lifeAddedSoFar++;
            this.lifes++;
            this.overlaysManager.setLifeCount(this.lifes - 1);
            this.overlaysManager.displayNewLife();
        }
    }
}

File: rust/src/game_manager.rs
use web_sys::CanvasRenderingContext2d;

use crate::{
    game_states::GameState,
    overlays::overlays_manager::{OverlayData, OverlayManager},
    sprites::{sprite::CanvasDimension, sprites_manager::SpriteManager},
    utils::{
        colors::Colors,
        config::Config,
        interval::Interval,
        javascript::{getCookie, setCookie},
    },
};

const GAME_OVER_STATE_TIMEOUT: u32 = 8000; // ms
const ADD_LIFE_WHEN_SCORED: u32 = 3000;
const ASTEROID_HIT_SCORE: u32 = 10;
const UFO_HIT_SCORE: u32 = 50;

pub struct GameManager {
    game_state: GameState,
    sprite_manager: SpriteManager,
    overlay_manager: OverlayManager,
    is_game_paused: bool,
    level: u32,
    lifes: u32,
    top_score: u32,
    score: u32,
    max_asteroids: u32,
    ufo_create_frequency: u32,
    ufo_shoot_frequency: u32,
    life_added_so_far: u32,
    game_over_interval: Interval,
    canvas: CanvasDimension,
    config: Config,
}

impl GameManager {
    pub fn new(canvas: CanvasDimension) -> GameManager {
        let top_score = getCookie();

        let mut sprite_manager = SpriteManager::new(canvas);
        sprite_manager.create_ufo(0);

        GameManager {
            game_state: GameState::Homescreen,
            sprite_manager,
            overlay_manager: OverlayManager::new(canvas),
            is_game_paused: false,
            level: 0,
            lifes: 0,
            top_score,
            score: 0,
            max_asteroids: 0,
            ufo_create_frequency: 0,
            ufo_shoot_frequency: 0,
            life_added_so_far: 0,
            game_over_interval: Interval::new(),
            canvas,
            config: Config::new(),
        }
    }

    pub fn key_pressed(&mut self, key: &str) {
        self.overlay_manager.key_pressed(key);

        if (self.game_state == GameState::Homescreen || self.game_state == GameState::GameOver)
            && key == "s"
        {
            self.start_game();
        }

        if self.game_state == GameState::Playing {
            self.sprite_manager.key_pressed(key);

            if key == "p" {
                self.is_game_paused = !self.is_game_paused;

                if self.is_game_paused {
                    self.sprite_manager.pause();
                } else {
                    self.sprite_manager.unpause();
                }
            }
        }

        if self.game_state == GameState::NextLevel && key == "s" {
            self.next_level();
        }

        if self.game_state == GameState::NextLife && key == "s" {
            self.start_level();
        }
    }

    pub fn key_released(&mut self, key: &str) {
        self.sprite_manager.key_released(key);
    }

    pub fn update(&mut self) {
        self.overlay_manager.update(self.game_state);

        if !self.is_game_paused {
            self.sprite_manager.update();

            if self.game_state == GameState::Playing {
                self.check_level();
                self.check_new_life();
            }
        }

        if self.game_over_interval.is_ellapsed() {
            self.game_over_interval.cancel();
            self.game_state = GameState::Homescreen;

            self.sprite_manager.reset();
            self.sprite_manager.create_ufo(0);
            self.sprite_manager
                .create_asteroids(self.config.game.asteroids_start_count);
        }
    }

    pub fn draw(&mut self, canvas: CanvasRenderingContext2d) {
        canvas.save();

        // Clear canvas
        canvas.set_stroke_style(&Colors::Edge.value().into());
        canvas.set_fill_style(&Colors::Background.value().into());
        canvas.set_line_width(1.5);

        canvas.fill_rect(0f64, 0f64, self.canvas.width, self.canvas.height);

        canvas.set_fill_style(&Colors::Edge.value().into());

        self.overlay_manager.draw_background(canvas.clone());

        self.sprite_manager.draw(canvas.clone());

        let overlay_data = OverlayData {
            game_state: self.game_state,
            is_game_paused: self.is_game_paused,
            top_score: self.top_score,
            score: self.get_score(),
            level: self.level,
            lifes: self.lifes,
            canvas: canvas.clone(),
        };

        self.overlay_manager.draw_foreground(overlay_data);

        canvas.restore();
    }

    fn start_game(&mut self) {
        self.game_over_interval.cancel();

        self.score = 0;
        self.level = 1;

        self.score = 0;
        self.level = 1;
        self.max_asteroids = self.config.game.asteroids_start_count;
        self.lifes = self.config.game.lifes_when_starting;

        self.overlay_manager.set_life_count(self.lifes - 1);

        self.life_added_so_far = 0;
        self.ufo_create_frequency = self.config.game.ufo_create_init_frequency;
        self.ufo_shoot_frequency = self.config.game.ufo_shoot_init_frequency;

        self.sprite_manager.reset();

        self.start_level();
    }

    fn start_level(&mut self) {
        self.game_state = GameState::Playing;

        self.sprite_manager.start_level(
            self.max_asteroids,
            self.ufo_create_frequency,
            self.ufo_shoot_frequency,
        );
    }

    pub fn next_level(&mut self) {
        self.level += 1;
        self.max_asteroids += self.config.game.asteroids_level_increment;
        self.ufo_create_frequency -= self.config.game.ufo_create_decrement_frequency;
        self.ufo_shoot_frequency -= self.config.game.ufo_shoot_decrement_frequency;

        if self.ufo_create_frequency < self.config.game.ufo_create_minimal_frequency {
            self.ufo_create_frequency = self.config.game.ufo_create_minimal_frequency;
        }

        if self.ufo_shoot_frequency < self.config.game.ufo_shoot_minimal_frequency {
            self.ufo_shoot_frequency = self.config.game.ufo_shoot_minimal_frequency;
        }

        self.start_level();
    }

    fn check_level(&mut self) {
        if self.sprite_manager.is_ship_active {
            if self.sprite_manager.get_asteroids_count() == 0 {
                self.game_state = GameState::NextLevel;
                self.score = self.get_score();
                self.sprite_manager.stop_level();
            }
        } else {
            self.lifes -= 1;
            self.score = self.get_score();
            self.sprite_manager.stop_level();

            if self.lifes == 0 {
                self.game_state = GameState::GameOver;
                self.game_over_interval.set(GAME_OVER_STATE_TIMEOUT);

                if self.score > self.top_score {
                    self.top_score = self.score;
                    setCookie(self.top_score);
                }
            } else {
                self.game_state = GameState::NextLife;
                self.overlay_manager.set_life_count(self.lifes - 1);
            }
        }
    }

    pub fn get_score(&self) -> u32 {
        self.score
            + self.sprite_manager.count_asteroids_hit * ASTEROID_HIT_SCORE
            + self.sprite_manager.count_ufo_hit * UFO_HIT_SCORE
    }

    pub fn check_new_life(&mut self) {
        if self.get_score() - self.life_added_so_far * ADD_LIFE_WHEN_SCORED >= ADD_LIFE_WHEN_SCORED
        {
            self.life_added_so_far += 1;
            self.lifes += 1;
            self.overlay_manager.set_life_count(self.lifes - 1);
            self.overlay_manager.dispaly_new_life();
        }
    }
}
